src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,337 @@
+    }
+
+    fn set_project_name_width(&self, width:f32) {
+        self.breadcrumbs_container.set_position(Vector3::new(HORIZONTAL_MARGIN + width,0.0,0.0));
+    }
+
+    fn select_breadcrumb(&self, index:usize) {
+        let current_index = self.current_index.get();
+        match index.cmp(&current_index) {
+            cmp::Ordering::Less => {
+                // If we have more crumbs after `index`, we will pop them.
+                let pop_amount    = current_index - index;
+                for _ in 0..pop_amount {
+                    self.breadcrumb_pop.emit(());
+                }
+            },
+            cmp::Ordering::Greater => {
+                for index in current_index..index {
+                    let info = self.breadcrumbs.borrow().get(index).map(|breadcrumb| {
+                        (breadcrumb.info.method_pointer.clone(),breadcrumb.info.expression_id)
+                    }).as_ref().cloned();
+                    if let Some((method_pointer,expression_id)) = info {
+                        self.breadcrumb_push.emit((method_pointer,expression_id));
+                    }
+                }
+            },
+            cmp::Ordering::Equal => ()
+        }
+        info!(self.logger,"Selecting breadcrumb #{index}");
+    }
+
+    fn push_breadcrumb(&self, method_pointer:&Rc<MethodPointer>, expression_id:&uuid::Uuid) {
+        let current_index = self.current_index.get();
+        let next_index    = current_index + 1;
+
+        let breadcrumb_exists =
+            self.breadcrumbs.borrow_mut().get(current_index).map(|breadcrumb| {
+                breadcrumb.info.expression_id
+            }).filter(|other_expression_id| other_expression_id == expression_id).is_some();
+
+        if breadcrumb_exists {
+            debug!(self.logger, "Entering an existing {method_pointer.name} breadcrumb.");
+            //TODO[dg]: Highlight breadcrumb.
+        } else {
+            debug!(self.logger, "Creating a new {method_pointer.name} breadcrumb.");
+            self.remove_breadcrumbs_history();
+            let breadcrumb       = Breadcrumb::new(&self.scene,method_pointer,expression_id);
+            let network          = &breadcrumb.frp.network;
+            let breadcrumb_index = next_index;
+            let model            = self.clone_ref();
+
+
+            // === User Interaction ===
+
+            frp::extend! { network
+                eval_ breadcrumb.frp.outputs.selected(model.select_breadcrumb(breadcrumb_index));
+            }
+
+
+            // === GUI Update ===
+
+            frp::extend! { network
+                eval self.project_name.frp.outputs.width((width) {
+                        model.set_project_name_width(*width)
+                });
+            }
+
+            info!(self.logger, "Pushing {breadcrumb.info.method_pointer.name} breadcrumb.");
+            breadcrumb.set_position(Vector3::new(self.width(), 0.0, 0.0));
+            breadcrumb.frp.fade_in.emit(());
```

Comment:
do not call it from model. You need to do it directly in FRP network. Doing this in model makes the whole logic VERY hard to follow :)